File: raffia-core/src/main/java/com/codepoetics/raffia/predicates/NumberPredicates.java
package com.codepoetics.raffia.predicates;

import com.codepoetics.raffia.mappers.Mapper;
import com.codepoetics.raffia.operations.ValuePredicate;

import java.math.BigDecimal;

public final class NumberPredicates {

  private NumberPredicates() {
  }

  public static ValuePredicate<BigDecimal> isEqualTo(int other) {
    return isEqualTo(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isEqualTo(long other) {
    return isEqualTo(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isEqualTo(double other) {
    return isEqualTo(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isEqualTo(String other) {
    return isEqualTo(new BigDecimal(other));
  }

  public static ValuePredicate<BigDecimal> isEqualTo(final BigDecimal other) {
    return new ValuePredicate<BigDecimal>() {
      @Override
      public boolean test(BigDecimal input) {
        return input.equals(other);
      }
    };
  }

  public static ValuePredicate<BigDecimal> isGreaterThan(int other) {
    return isGreaterThan(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isGreaterThan(long other) {
    return isGreaterThan(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isGreaterThan(double other) {
    return isGreaterThan(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isGreaterThan(String other) {
    return isGreaterThan(new BigDecimal(other));
  }
  
  public static ValuePredicate<BigDecimal> isGreaterThan(final BigDecimal other) {
    return new ValuePredicate<BigDecimal>() {
      @Override
      public boolean test(BigDecimal input) {
        return input.compareTo(other) > 0;
      }
    };
  }

  public static ValuePredicate<BigDecimal> isLessThan(int other) {
    return isLessThan(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isLessThan(long other) {
    return isLessThan(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isLessThan(double other) {
    return isLessThan(BigDecimal.valueOf(other));
  }
  
  public static ValuePredicate<BigDecimal> isLessThan(String other) {
    return isLessThan(new BigDecimal(other));
  }
  
  public static ValuePredicate<BigDecimal> isLessThan(final BigDecimal other) {
    return new ValuePredicate<BigDecimal>() {
      @Override
      public boolean test(BigDecimal input) {
        return input.compareTo(other) < 0;
      }
    };
  }

  public static ValuePredicate<BigDecimal> isGreaterThanOrEqualTo(int other) {
    return isGreaterThanOrEqualTo(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isGreaterThanOrEqualTo(long other) {
    return isGreaterThanOrEqualTo(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isGreaterThanOrEqualTo(double other) {
    return isGreaterThanOrEqualTo(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isGreaterThanOrEqualTo(String other) {
    return isGreaterThanOrEqualTo(new BigDecimal(other));
  }
  
  public static ValuePredicate<BigDecimal> isGreaterThanOrEqualTo(final BigDecimal other) {
    return new ValuePredicate<BigDecimal>() {
      @Override
      public boolean test(BigDecimal input) {
        return input.compareTo(other) >= 0;
      }
    };
  }

  public static ValuePredicate<BigDecimal> isLessThanOrEqualTo(int other) {
    return isLessThanOrEqualTo(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isLessThanOrEqualTo(long other) {
    return isLessThanOrEqualTo(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isLessThanOrEqualTo(double other) {
    return isLessThanOrEqualTo(BigDecimal.valueOf(other));
  }

  public static ValuePredicate<BigDecimal> isLessThanOrEqualTo(String other) {
    return isLessThanOrEqualTo(new BigDecimal(other));
  }
  
  public static ValuePredicate<BigDecimal> isLessThanOrEqualTo(final BigDecimal other) {
    return new ValuePredicate<BigDecimal>() {
      @Override
      public boolean test(BigDecimal input) {
        return input.compareTo(other) <= 0;
      }
    };
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/predicates/NumberPredicates.kt
package com.codepoetics.raffia.predicates

import java.math.BigDecimal

object NumberPredicates {

    @JvmStatic
    fun isEqualTo(other: Int): ValuePredicate<BigDecimal> = isEqualTo(BigDecimal.valueOf(other.toLong()))

    @JvmStatic
    fun isEqualTo(other: Long): ValuePredicate<BigDecimal> = isEqualTo(BigDecimal.valueOf(other))

    @JvmStatic
    fun isEqualTo(other: Double): ValuePredicate<BigDecimal> = isEqualTo(BigDecimal.valueOf(other))

    @JvmStatic
    fun isEqualTo(other: String): ValuePredicate<BigDecimal> = isEqualTo(BigDecimal(other))

    @JvmStatic
    fun isEqualTo(other: BigDecimal): ValuePredicate<BigDecimal> = { input -> input == other }

    @JvmStatic
    fun isGreaterThan(other: Int): ValuePredicate<BigDecimal> = isGreaterThan(BigDecimal.valueOf(other.toLong()))

    @JvmStatic
    fun isGreaterThan(other: Long): ValuePredicate<BigDecimal> = isGreaterThan(BigDecimal.valueOf(other))

    @JvmStatic
    fun isGreaterThan(other: Double): ValuePredicate<BigDecimal> = isGreaterThan(BigDecimal.valueOf(other))

    @JvmStatic
    fun isGreaterThan(other: String): ValuePredicate<BigDecimal> = isGreaterThan(BigDecimal(other))

    @JvmStatic
    fun isGreaterThan(other: BigDecimal): ValuePredicate<BigDecimal> = { input -> input > other }

    @JvmStatic
    fun isLessThan(other: Int): ValuePredicate<BigDecimal> = isLessThan(BigDecimal.valueOf(other.toLong()))

    @JvmStatic
    fun isLessThan(other: Long): ValuePredicate<BigDecimal> = isLessThan(BigDecimal.valueOf(other))

    @JvmStatic
    fun isLessThan(other: Double): ValuePredicate<BigDecimal> = isLessThan(BigDecimal.valueOf(other))

    @JvmStatic
    fun isLessThan(other: String): ValuePredicate<BigDecimal> = isLessThan(BigDecimal(other))

    @JvmStatic
    fun isLessThan(other: BigDecimal): ValuePredicate<BigDecimal> = { input -> input < other }

    @JvmStatic
    fun isGreaterThanOrEqualTo(other: Int): ValuePredicate<BigDecimal> = isGreaterThanOrEqualTo(BigDecimal.valueOf(other.toLong()))

    @JvmStatic
    fun isGreaterThanOrEqualTo(other: Long): ValuePredicate<BigDecimal> = isGreaterThanOrEqualTo(BigDecimal.valueOf(other))

    @JvmStatic
    fun isGreaterThanOrEqualTo(other: Double): ValuePredicate<BigDecimal> = isGreaterThanOrEqualTo(BigDecimal.valueOf(other))

    @JvmStatic
    fun isGreaterThanOrEqualTo(other: String): ValuePredicate<BigDecimal> = isGreaterThanOrEqualTo(BigDecimal(other))

    @JvmStatic
    fun isGreaterThanOrEqualTo(other: BigDecimal): ValuePredicate<BigDecimal> = { input -> input >= other }

    @JvmStatic
    fun isLessThanOrEqualTo(other: Int): ValuePredicate<BigDecimal> = isLessThanOrEqualTo(BigDecimal.valueOf(other.toLong()))

    @JvmStatic
    fun isLessThanOrEqualTo(other: Long): ValuePredicate<BigDecimal> = isLessThanOrEqualTo(BigDecimal.valueOf(other))

    @JvmStatic
    fun isLessThanOrEqualTo(other: Double): ValuePredicate<BigDecimal> = isLessThanOrEqualTo(BigDecimal.valueOf(other))

    @JvmStatic
    fun isLessThanOrEqualTo(other: String): ValuePredicate<BigDecimal> = isLessThanOrEqualTo(BigDecimal(other))

    @JvmStatic
    fun isLessThanOrEqualTo(other: BigDecimal): ValuePredicate<BigDecimal> = { input -> input <= other }
}
